content/articles/2017-12-workflow-vs-script.md

```diff
@@ -0,0 +1,126 @@
+
+## Self-sufficiency
+
+I suggest organizing data analyses into *projects*: a folder on your computer that holds all the files relevant to a particular piece of work. Any resident R script should assume that it will be run from a fresh R process with working directory set to the directory where this project lives. This convention guarantees that the project can be moved around on your computer or onto other computers and will still "just work". I argue that this is the only practical convention that creates reliable behavior across different computers or users and over time.
+
+## What's wrong with `setwd()`?
+
+I run alot of student code in [STAT 545](http://stat545.com) and, at the start, I see alot of R scripts that look like this:
+
+``` r
+setwd("/Users/jenny/cuddly_broccoli/verbose_funicular/foofy/data")
+df <- read.delim("raw_foofy_data.csv")
+p <- ggplot(df, aes(x, y)) + geom_point()
+ggsave("../figs/foofy_scatterplot.png")
+```
+
+The chance of the `setwd()` command having the desired effect -- making the file paths work -- for anyone besides its author is 0%. It's also unlikely to work for the author one or two years or computers from now. The project is not self-contained and is not portable. To recreate and perhaps extend this plot, the lucky recipient will need to hand edit one or more paths to reflect where the project has landed on their machine. When you do this for the 73rd time in 2 days, while marking an assignment, you start to fantasize about lighting the perpetrator's computer on fire.
+
+This use of `setwd()` is also highly suggestive that the useR does all of their work in one R process and manually switches gears when they shift from one project to another. That sort of workflow makes it unpleasant to work on more than one project at a time and also makes it easy for work done on one project to accidentially leak into subsequent work on another (e.g., objects, loaded packages, session options).
+
+## Use the [here package](https://CRAN.R-project.org/package=here) and Projects
+
+How can you avoid `setwd()` at the top of every script?
+
+  * Organize each logical project into a folder on your computer.
+  * Make sure the top-level folder advertises itself as such: I HOLD A PROJECT! This can be as simple as having an empty file named `.here`. Or, if you use RStudio and/or Git, those both leave characteristic files behind that will get the job done.
+  * Use the `here()` function from the [here package](https://CRAN.R-project.org/package=here) to build the path when you read or write a file. Create paths relative to the top-level directory.
+  
+Whenever you work on this project, launch the R process from the project's top-level directory. To continue our example, make sure R is launched in the `foofy` directory, wherever that may be. Now the code looks like so:
```

Comment:
I don't know where this should go, but in my `.profile` I have:

```
alias R='R --no-save --no-restore-data --quiet'
```

which seems like a good idea if you're going to be running R from the console. Maybe a footnote? Also needs mention somewhere that R inherits its working direction from the terminal's working directory.